beeline/src/java/org/apache/hive/beeline/JSONOutputFormat.java

```diff
@@ -0,0 +1,129 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * This source file is based on code taken from SQLLine 1.9
+ * See SQLLine notice in LICENSE
+ */
+package org.apache.hive.beeline;
+
+import java.sql.SQLException;
+import java.sql.Types;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+
+import com.fasterxml.jackson.core.JsonEncoding;
+import com.fasterxml.jackson.core.JsonFactory;
+import com.fasterxml.jackson.core.JsonGenerator;
+
+/**
+ * OutputFormat for standard JSON format.
+ * 
+ */ 
+public class JSONOutputFormat extends AbstractOutputFormat {
+  protected final BeeLine beeLine;
+  protected JsonGenerator generator;
+  
+
+  /**
+   * @param beeLine
+   */
+  JSONOutputFormat(BeeLine beeLine){ 
+    this.beeLine = beeLine;
+    ByteArrayOutputStream buf = new ByteArrayOutputStream();
+    try {
+      this.generator = new JsonFactory().createGenerator(buf, JsonEncoding.UTF8);
+    } catch(IOException e) {
+      beeLine.handleException(e);
+    }
+  }
+
+  @Override 
+  void printHeader(Rows.Row header) {
+    try {
+      generator.writeStartObject();
+      generator.writeArrayFieldStart("resultset");
+    } catch(IOException e) {
+      beeLine.handleException(e);
+    }
+  }
+
+  @Override 
+  void printFooter(Rows.Row header) {
+    try {
+      generator.writeEndArray();
+      generator.writeEndObject();
+      generator.flush();
+      String out = ((ByteArrayOutputStream) generator.getOutputTarget()).toString("UTF-8");
+      beeLine.output(out);
+    } catch(IOException e) {
+      beeLine.handleException(e);
+    }
+  }
+
+  @Override 
+  void printRow(Rows rows, Rows.Row header, Rows.Row row) {
+    String[] head = header.values;
+    String[] vals = row.values;
+    
+    try {
+      int colCount = rows.rsMeta.getColumnCount();
+      boolean objStartFlag = true;
+
+      for (int i = 0; (i < head.length) && (i < vals.length); i++) {
+        if (objStartFlag) {
+          generator.writeStartObject();
+          objStartFlag = false;
+        }
+
+        generator.writeFieldName(head[i]);
+        switch(rows.rsMeta.getColumnType(i+1)) {
+          case Types.TINYINT:
+          case Types.SMALLINT:
+          case Types.INTEGER:
+          case Types.BIGINT:
+          case Types.REAL:
+          case Types.FLOAT:
+          case Types.DOUBLE:
+          case Types.DECIMAL:
+          case Types.NUMERIC:
+          case Types.ROWID:
+            generator.writeNumber(vals[i]);
+            break;
+          case Types.NULL:
+            generator.writeNull();
+            break;
+          case Types.BOOLEAN:
+            generator.writeBoolean(Boolean.parseBoolean(vals[i]));
+            break;
+          default:
```

Comment:
Need a type for BINARY data.  Make it spit it out as Base-64 to be text-friendly.

Reply:
So `Rows.java` already does this conversion 
```
else if (o instanceof byte[]) { 
  value = convertBinaryArray ? new String((byte[])o) : Arrays.toString((byte[])o);
}
```
I'm not 100% sure what the expected behavior here in the output format would be then? Currently it falls into the default case and prints the already converted `byte[]`

Reply:
Thinking this through more `Arrays.toString(byte[] a)` will convert it to some string that looks like `[16, 34, 67]`, meaning the resulting JSON will be `{..."binary_field":"[16,34,67]"...}`.

`JSONOutputFormat` does not have access to the underlying resultset, just the values that have already been `toString()`ed. Is the correct behavior then to convert back to `byte[]` and then `Base64.encode()`?

Reply:
Ya, this is not a particularly helpful format.  The default in JSON is Base64 for binary data, so you should use that and not rely on this default

Reply:
https://fasterxml.github.io/jackson-core/javadoc/2.8/com/fasterxml/jackson/core/JsonGenerator.html#writeBinary(byte[])

Reply:
OK.  I follow now.  Sorry I didn't read your comment close enough. :)

So, I think for now, what you should so is... 

Create a new `case` for BINARY data and then check that that this is set to "true".  Throw an Exception otherwise.

```
beeLine.getOpts().getConvertBinaryArrayToString();
```

At least this way, we can have a separate ticket about making this base-64 instead of the current implementation.